### PR TITLE
Persist columns

### LINKED ISSUE #469: Update JSON (de)serialization for Catalog Objects
Schema and IndexSchema have been updated and need the JSON serialization/deserialization functions updated to reflect the recursion down into AbstractExpression.

### LINKED ISSUE #470: Persist column information into pg_attribute
Finish logic to persist column metadata into pg_attribute.

Blocked on #386.

### PR DESCRIPTION
**Pull request by jrolli**
This PR makes the changes necessary to persist the columns of both tables and indexes by inserting (and hence logging) them into `pg_attribute`.  The design has changed due to constraints on expression ownership so we are no longer using the `ADBIN` column and instead only store the materialized expressions in the corresponding schema.

In order to fully support this, I added the code to properly serialize/deserialize the (index) schema objects.

This PR also fixes the deallocation approach for the database catalog to rely on per table/schema abort deletions rather than trying to reuse the final teardown logic.

The PR also includes the bug fixes to transient value serialization (#480) but does not include the additional test in that PR.

Fixes #469 
Fixes #470

### CODE REVIEW

File: src/catalog/database_catalog.cpp [hunk 1]
```diff
@@ -368,8 +369,8 @@ bool DatabaseCatalog::CreateColumn(transaction::TransactionContext *const txn, c
   *len_entry = (col.Type() == type::TypeId::VARCHAR || col.Type() == type::TypeId::VARBINARY) ? col.MaxVarlenSize()
                                                                                               : col.AttrSize();
   *notnull_entry = !col.Nullable();
-  *dbin_entry = reinterpret_cast<uintptr_t>(
-      nullptr);  // TODO(Matt) implement deep copy of the abstract expression once #386 is in
+  // TODO(John) implement deep copy of the abstract expression and store the pointer once #386 is in
+  *dbin_entry = nullptr;
```
Comment (GustavoAngulo) [inline src/catalog/database_catalog.cpp]: I'm a bit confused, your PR description said we are no longer using the ADBIN column. Do you mean we should deprecate it, or just always set it to nullptr like done here.

File: src/catalog/database_catalog.cpp [hunk 2]
```diff
@@ -550,6 +544,13 @@ bool DatabaseCatalog::DeleteColumns(transaction::TransactionContext *const txn,
         reinterpret_cast<const uint32_t *const>(pr->AccessWithNullCheck(table_pm[ATTNUM_COL_OID]));
     TERRIER_ASSERT(col_name != nullptr, "OID shouldn't be NULL.");
 
+    auto *stored_expression =
+        *reinterpret_cast<parser::AbstractExpression **>(pr->AccessForceNotNull(table_pm[ADBIN_COL_OID]));
```
Comment (GustavoAngulo) [inline src/catalog/database_catalog.cpp]: You're using `ADBIN_COL_OID` here, but it's not among the oids in `table_oids` above.

File: src/catalog/database_catalog.cpp [hunk 3]
```diff
@@ -759,6 +749,8 @@ table_oid_t DatabaseCatalog::GetTableOid(transaction::TransactionContext *const
 
 bool DatabaseCatalog::SetTablePointer(transaction::TransactionContext *const txn, const table_oid_t table,
                                       const storage::SqlTable *const table_ptr) {
+  auto *tm = txn->GetTransactionManager();
+  txn->RegisterAbortAction([=]() { tm->DeferAction([=]() { delete table_ptr; }); });
```
Comment (GustavoAngulo) [inline src/catalog/database_catalog.cpp]: Can you add a comment explaining why we defer this action here, and why the catalog does it? I would expect the caller to check the return value of `SetTablePointer` and clean up the table it created if the call failed.

File: src/catalog/database_catalog.cpp [hunk 4]
```diff
@@ -1027,6 +1019,8 @@ bool DatabaseCatalog::SetClassPointer(transaction::TransactionContext *const txn
 
 bool DatabaseCatalog::SetIndexPointer(transaction::TransactionContext *const txn, const index_oid_t index,
                                       const storage::index::Index *const index_ptr) {
+  auto *tm = txn->GetTransactionManager();
+  txn->RegisterAbortAction([=]() { tm->DeferAction([=]() { delete index_ptr; }); });
```
Comment (GustavoAngulo) [inline src/catalog/database_catalog.cpp]: Same comment as `SetTablePointer`.

File: src/include/catalog/index_schema.h [hunk 5]
```diff
@@ -107,6 +106,38 @@ class IndexSchema {
...
+     * Deserializes a column
+     * @param j serialized column
+     */
+    void FromJson(const nlohmann::json &j) {
+      name_ = j.at("name").get<std::string>();
+      SetTypeId(j.at("type").get<type::TypeId>());
+      SetMaxVarlenSize(j.at("max_varlen_size").get<uint16_t>());
+      SetNullable(j.at("nullable").get<bool>());
+      SetOid(j.at("oid").get<indexkeycol_oid_t>());
+      definition_ = j.at("definition").get<std::shared_ptr<parser::AbstractExpression>>();
```
Comment (GustavoAngulo) [inline src/include/catalog/index_schema.h]: This line won't work, use the `parser::expression::DeserializeExpression` method.

File: src/include/catalog/schema.h [hunk 6]
```diff
@@ -145,6 +148,7 @@ class Schema {
       max_varlen_size_ = j.at("max_varlen_size").get<uint16_t>();
       nullable_ = j.at("nullable").get<bool>();
       oid_ = j.at("oid").get<col_oid_t>();
+      default_value_ = j.at("default_value").get<std::shared_ptr<parser::AbstractExpression>>();
```
Comment (GustavoAngulo) [inline src/include/catalog/schema.h]: Use `parser::expression::DeserializeExpression`

File: src/catalog/database_catalog.cpp [hunk 7]
```diff
@@ -550,6 +544,13 @@ bool DatabaseCatalog::DeleteColumns(transaction::TransactionContext *const txn,
         reinterpret_cast<const uint32_t *const>(pr->AccessWithNullCheck(table_pm[ATTNUM_COL_OID]));
     TERRIER_ASSERT(col_name != nullptr, "OID shouldn't be NULL.");
 
+    auto *stored_expression =
+        *reinterpret_cast<parser::AbstractExpression **>(pr->AccessForceNotNull(table_pm[ADBIN_COL_OID]));
+
+    if (stored_expression != nullptr) {
+      txn->RegisterCommitAction([=]() { delete stored_expression; });
```
Comment (GustavoAngulo) [inline src/catalog/database_catalog.cpp]: Who cleans `stored_expression` up in the case of an abort? We should probably register a Commit and Abort action.

File: src/include/catalog/database_catalog.h [hunk 8]
```diff
@@ -265,7 +269,7 @@ class DatabaseCatalog {
    * @return the column from pg_attribute
    */
   template <typename Column, typename ClassOid, typename ColOid>
-  std::vector<std::unique_ptr<Column>> GetColumns(transaction::TransactionContext *txn, ClassOid class_oid);
+  std::vector<Column> GetColumns(transaction::TransactionContext *txn, ClassOid class_oid);
   // TODO(Matt): make this return stack object
```
Comment (GustavoAngulo) [inline src/include/catalog/database_catalog.h]: nit: You addressed this TODO

File: src/include/catalog/database_catalog.h [hunk 9]
```diff
@@ -253,8 +257,8 @@ class DatabaseCatalog {
    * @param class_oid oid of table or index
    * @return the column from pg_attribute
    */
-  template <typename Column, typename ClassOid, typename ColOid>
-  std::unique_ptr<Column> GetColumn(transaction::TransactionContext *txn, ClassOid class_oid, ColOid col_oid);
+  // template <typename Column, typename ClassOid, typename ColOid>
+  // Column GetColumn(transaction::TransactionContext *txn, ClassOid class_oid, ColOid col_oid);
```
Comment (GustavoAngulo) [inline src/include/catalog/database_catalog.h]: Why is this commented out?

Comment (jrolli) [reply src/include/catalog/database_catalog.h]: As best as I could tell, there is no actual use-case for materializing a single column so I commented it out during the update.  I've now completely removed it.

File: src/catalog/database_catalog.cpp [hunk 10]
```diff
@@ -759,6 +697,11 @@ table_oid_t DatabaseCatalog::GetTableOid(transaction::TransactionContext *const
 
 bool DatabaseCatalog::SetTablePointer(transaction::TransactionContext *const txn, const table_oid_t table,
                                       const storage::SqlTable *const table_ptr) {
+  auto *tm = txn->GetTransactionManager();
```
Comment (mbutrovich) [inline src/catalog/database_catalog.cpp]: Style nit: refer `txn_manager` over `tm`. I suspect this is actually all over the catalog.

File: src/include/catalog/index_schema.h [hunk 11]
```diff
@@ -116,8 +161,9 @@ class IndexSchema {
     std::string name_;
     indexkeycol_oid_t oid_;
     uint32_t packed_type_;
-    common::ManagedPointer<const parser::AbstractExpression> definition_;
-    std::string serialized_expression_;
+
+    // TODO(John) this should go back to being a raw pointer once #386 is in
```
Comment (mbutrovich) [inline src/include/catalog/index_schema.h]: Outdated TODO. Not sure if we know what PR to reference at this point.

File: src/catalog/database_catalog.cpp [hunk 12]
```diff
@@ -1516,12 +1458,32 @@ bool DatabaseCatalog::CreateTableEntry(transaction::TransactionContext *const tx
...
+  *(reinterpret_cast<namespace_oid_t *>(index_pr->AccessForceNotNull(0))) = ns_oid;
   const auto result UNUSED_ATTRIBUTE = classes_namespace_index_->Insert(txn, *index_pr, tuple_slot);
   TERRIER_ASSERT(result, "Insertion into non-unique namespace index failed.");
 
   delete[] index_buffer;
 
+  // Write the col oids into a new Schema object
+  col_oid_t curr_col_oid(1);
+  for (auto &col : schema.GetColumns()) {
+    auto result = CreateColumn(txn, table_oid, curr_col_oid++, col);
```
Comment (mbutrovich) [inline src/catalog/database_catalog.cpp]: Shadows another `result` variable.

File: src/catalog/database_catalog.cpp [hunk 13]
```diff
@@ -1310,6 +1231,27 @@ bool DatabaseCatalog::CreateIndexEntry(transaction::TransactionContext *const tx
   // Free the buffer, we are finally done
   delete[] index_buffer;
 
+  // Write the col oids into a new Schema object
+  indexkeycol_oid_t curr_col_oid(1);
+  for (auto &col : schema.GetColumns()) {
+    auto result = CreateColumn(txn, index_oid, curr_col_oid++, col);
```
Comment (mbutrovich) [inline src/catalog/database_catalog.cpp]: Shadows another `result` variable.

File: src/include/catalog/schema.h [hunk 14]
```diff
@@ -154,7 +176,9 @@ class Schema {
     uint16_t max_varlen_size_;
     bool nullable_;
     col_oid_t oid_;
-    common::ManagedPointer<const parser::AbstractExpression> default_value_;
+
+    // TODO(John) this should go back to being a raw pointer once #386 is in
```
Comment (mbutrovich) [inline src/include/catalog/schema.h]: Same issue with an outdated PR reference.

File: src/include/catalog/index_schema.h (hunk 11 above)
Comment (jrolli) [reply src/include/catalog/index_schema.h]: Changed to reference the new issue for smart pointers (#489).

File: src/include/catalog/schema.h (hunk 14 above)
Comment (jrolli) [reply src/include/catalog/schema.h]: Changed to reference the new issue for smart pointers (#489).

**Comment by jrolli** [comment]
I addressed the lack of testing for `DeleteColumns` by making the deletions in `pg_class` cascade into `pg_attributes` which should address that oversight.

The current shortcomings for the catalog testing is testing behavior under unexpected, but non-fatal, API calls.

There are also additional behavior tests we will want to add as we implement features such as cascading deletes for namespaces and tables, but overall I think we are testing most of the known behavior.

**Pull request merged**